MaxText/standalone_checkpointer.py

```diff
@@ -93,22 +92,53 @@ def checkpoint_loop(config, state=None):
         ckpt_write_time.append([jax.process_index(), checkpoint_write_time])
         if jax.process_index() == 0:
           max_logging.log(f"STANDALONE CHECKPOINTER : Checkpoint saved in {end_time - start_time} ,step {step}, on host 0")
+    if jax.process_index() == 0:
+      elapsed_time = datetime.datetime.now() - start_time
+      time_to_wait = config.step_time_seconds - elapsed_time.total_seconds()
+      if time_to_wait > 0:
+        max_logging.log(f"Waiting {time_to_wait} seconds to reach step time of {config.step_time_seconds} seconds for step {step}")
           
   if config.gcs_metrics_bucket:
-    max_logging.log(f"Uploading metrics to GCS bucket {config.gcs_metrics_bucket} on host {jax.process_index()}")
+    max_logging.log(f"Uploading write metrics to GCS bucket {config.gcs_metrics_bucket} on host {jax.process_index()}")
     
     base_name = f"{jax.process_index()}.csv"
     storage_utils.upload_csv(
         config.gcs_metrics_bucket, 
-        os.path.join(config.run_name, CHECKPOINT_RESTORE_TIME_DIRECTORY, base_name), 
-        ckpt_read_time
+        os.path.join(config.run_name, CHECKPOINT_WRITE_TIME_DIRECTORY, base_name), 
+        ckpt_write_time
     )
+    max_logging.log(f"Finished uploading write metrics to GCS bucket {config.gcs_metrics_bucket} on host {jax.process_index()}  for run {config.run_name}")
+
+  for step in np.arange(start_step, config.steps):
+    if checkpoint_manager is not None:
+      start_time = datetime.datetime.now()
+      # A barrier to sync all hosts before starting to save checkpoint
+      jax.experimental.multihost_utils.sync_global_devices("Barrier before restore")
+      try:
+        checkpoint_manager.restore(
+              step,
+              args=orbax.checkpoint.args.Composite(items=orbax.checkpoint.args.PyTreeRestore(item=unboxed_abstract_state)),
+            )
+      except FileNotFoundError:
+        # No checkpoint was found for the step, presumably because one was not produced for the step. Continue on.
+        continue
+      jax.block_until_ready(state)
```

Comment:
There is no "state" here, right? Shall we use the barrier, "sync_global_devices"?

Reply:
Updated checkpoint_manager.restore to restore to state, does this look good now?

Reply:
LGTM. (the variable scoping in Python is really interesting that you can access state like this).

Reply:
State is actually initially declared at the top of the function when it restores the initial checkpoint, although I think you could still do this even if not